Source/Core/Duality/Resources/RectAtlas.cs

```diff
@@ -0,0 +1,497 @@
+		/// <param name="item">The rect to insert</param>
+		/// <exception cref="ArgumentOutOfRangeException">
+		/// The given index was negative or greater than the number of rects in the atlas.
+		/// </exception>
+		public void Insert(int index, Rect item)
+		{
+			if (index < 0)
+				throw new ArgumentOutOfRangeException("index", "The insertion index cannot be less than 0");
+			if (index > this.items.Count)
+				throw new ArgumentOutOfRangeException("index", "The insertion index was greater than the size number of rects in the atlas");
+
+			if (this.tags != null)
+			{
+				// Adjust tagged indices to account for the insertion of this item
+				foreach (List<int> indexList in this.tags.Values)
+				{
+					for (int i = 0; i < indexList.Count; i++)
+					{
+						// Index after rect being added - update it to new index
+						if (indexList[i] >= index)
+						{
+							indexList[i]++;
+						}
+					}
+				}
+			}
+
+			this.items.Insert(index, new RectAtlasItem { Rect = item });
+		}
+
+		/// <summary>
+		/// Removes the rect from that atlas at the given index
+		/// </summary>
+		/// <param name="index">The index of the rect to remove</param>
+		/// <exception cref="ArgumentOutOfRangeException">
+		/// The given index was negative or greater than the number of rects in the atlas.
+		/// </exception>
+		public void RemoveAt(int index)
+		{
+			if (index < 0)
+				throw new ArgumentOutOfRangeException("index", "The insertion index cannot be less than 0");
+			if (index >= this.items.Count)
+				throw new ArgumentOutOfRangeException("index", "The insertion index was greater than the size number of rects in the atlas");
+
+			if (this.tags != null)
+			{
+				List<string> emptyKeys = new List<string>();
+
+				// Adjust tagged indices to account for the removal of this item
+				foreach (var kvp in this.tags)
+				{
+					List<int> indexList = kvp.Value;
+					for (int i = 0; i < indexList.Count; i++)
+					{
+						// Index of rect being removed - remove from tag list
+						if (indexList[i] == index)
+						{
+							indexList.RemoveAt(i);
+							i--;
+						}
+						// Index after rect being remove - update it to new index
+						else if (indexList[i] > index)
+						{
+							indexList[i]--;
+						}
+					}
+
+					if (indexList.Count == 0)
+						emptyKeys.Add(kvp.Key);
+				}
+
+				// Remove any list of tagged indices that are now empty
+				foreach (string emptyKey in emptyKeys)
+					this.tags.Remove(emptyKey);
+			}
+
+			this.items.RemoveAt(index);
+		}
+
+		/// <summary>
+		/// Returns all rects with the given tag or an empty enumerable if no rects
+		/// with the given tag exist. The returned rects will be in index order.
+		/// </summary>
+		/// <param name="tag">The tag of the rects to look for</param>
+		/// <returns>
+		/// The rects with the given tag or an empty enumerable if no rects with the given tag exist.
+		/// </returns>
+		public IEnumerable<Rect> GetTaggedRects(string tag)
```

Comment:
Avoid using LINQ and `IEnumerable<T>` here to avoid garbage and improve performance. Consider the following method syntax instead: `void GetTaggedItems(string tag, List<RectAtlasItem> items)`.